File: 1007_minimum_domino_rotations_for_equal_row/min_domino_rotations.java
class Solution {
    private int[] a;
    private int[] b;

    // Time: O(n)
    // Space: O(1)
    public int minDominoRotations(int[] A, int[] B) {
        a = A;
        b = B;
        int ops = check(a[0]);
        if (ops != -1 || a[0] == b[0]) {
            return ops;
        }
        return check(b[0]);
    }

    private int check(int val) {
        int x = 0, y = 0;
        for (int i = 0; i < a.length; i++) {
            if (a[i] != val && b[i] != val) {
                return -1;
            } else if (a[i] != val) {
                x++;
            } else if (b[i] != val) {
                y++;
            }
        }
        return Math.min(x, y);
    }
}

File: 1007_minimum_domino_rotations_for_equal_row/min_domino_rotations.py
from typing import List


def min_domino_rotations(A: List[int], B: List[int]) -> int:
    """
    Time: O(m + n)
    Space: O(1)
    """
    def check(x):
        a = b = 0
        for i in range(len(A)):
            if A[i] != x and B[i] != x:
                return -1
            elif A[i] != x:
                a += 1
            elif B[i] != x:
                b += 1
        return min(a, b)
    ops = check(A[0])
    if ops != -1 or A[0] == B[0]:
        return ops
    return check(B[0])
